### PR TITLE
Initial implementation of context aware alert analysis

### PR DESCRIPTION
**Pull request by songkant-aws**
### Description
Initial context aware alert analysis implementation in assistant dashboard by leveraging existing IncontextInsight 'GeneratePopover' component

### Issues Resolved
[[Issue #214 ]](https://github.com/opensearch-project/dashboards-assistant/issues/214)

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass, including unit test, integration test.
- [x] New functionality has user manual doc added.
- [x] Commits are signed per the DCO using --signoff.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by wanglam** [comment]
@songkant-aws Is this PR target for 2.16?

File: public/components/incontext_insight/index.tsx [hunk 1]
```diff
@@ -195,19 +348,36 @@ export const IncontextInsight = ({ children }: IncontextInsightProps) => {
     </>
   );
 
-  const ChatPopoverBody: React.FC<{}> = ({}) => (
-    <EuiFlexGroup>
+  const [userQuestion, setUserQuestion] = useState('');
```
Comment (wanglam) [inline public/components/incontext_insight/index.tsx:351]: Shall we move this `useState` inside `ChatPopoverBody`? Seems this state only been used in this component.

File: public/components/incontext_insight/index.tsx [hunk 2]
```diff
@@ -174,17 +258,86 @@ export const IncontextInsight = ({ children }: IncontextInsightProps) => {
     </EuiPopoverFooter>
   );
 
-  const GeneratePopoverBody: React.FC<{}> = ({}) => (
-    <EuiButton onClick={() => toasts.addDanger('To be implemented...')}>Generate summary</EuiButton>
-  );
+  const GeneratePopoverBody: React.FC<{ incontextInsight: IncontextInsightInput }> = ({
```
Comment (wanglam) [inline public/components/incontext_insight/index.tsx:261]: Can we separate `GeneratePopoverBody` outside as an isolate component? It take about 120 lines. Prefer move `onGenerateSummary`, `onChatContinuation` and `isLlmResponded`, `isLoading` state in the same component too.

**Comment by songkant-aws** [comment]
> @songkant-aws Is this PR target for 2.16?

Yes, the PR is targeting for 2.16

File: common/types/chat_saved_object_attributes.ts [hunk 3]
```diff
@@ -43,8 +45,11 @@ export interface IInput {
   content: string;
   context?: {
     appId?: string;
+    content?: string;
+    datasourceId?: string;
   };
   messageId?: string;
+  promptPrefix?: AssistantRole;
```
[the remark this answers is not in the document]
Comment (songkant-aws) [reply common/types/chat_saved_object_attributes.ts]: I make prompPrefix string for IInput structure

File: server/services/chat/olly_chat_service.ts [hunk 4]
```diff
@@ -9,11 +9,33 @@ import { IMessage, IInput } from '../../../common/types/chat_saved_object_attrib
 import { ChatService } from './chat_service';
 import { ML_COMMONS_BASE_API, ROOT_AGENT_CONFIG_ID } from '../../utils/constants';
 
+export enum AssistantRole {
```
[the remark this answers is not in the document]
Comment (songkant-aws) [reply server/services/chat/olly_chat_service.ts]: Moved all of AssistantRole logic to public/types.ts

File: server/routes/chat_routes.ts [hunk 5]
```diff
@@ -232,6 +235,17 @@ export function registerChatRoutes(router: IRouter, routeOptions: RoutesOptions)
...
         }
 
+        resultPayload.messages
+          .filter((message) => message.type === 'input')
+          .forEach((msg) => {
+            // hide additional conetxt to how was it generated
+            const index = msg.content.indexOf('answer question:');
+            const len = 'answer question:'.length;
+            if (index !== -1) {
+              msg.content = msg.content.substring(index + len);
```
[the remark this answers is not in the document]
Comment (songkant-aws) [reply server/routes/chat_routes.ts:245]: Actually, the logic of this part is not removing 'answer question:' substring from the original string. I believe it just needs question part as part of message input. So "blabla. answer question: summarization question" message content will strip away substring until the index after 'answer question:'. The ' summarization question' is kept. 

The mechanism here is probably weird but I'm not aware there is other convenient way to replace prompt of backend agent workflow. May need to discuss about it with Hailong a bit.

File: public/components/incontext_insight/index.tsx (hunk 1 above)
Comment (songkant-aws) [reply public/components/incontext_insight/index.tsx]: Done

File: public/components/incontext_insight/index.tsx (hunk 2 above)
Comment (songkant-aws) [reply public/components/incontext_insight/index.tsx]: Now it's an independent component in a different file - generate_popover_body.tsx

File: public/components/incontext_insight/generate_popover_body.tsx [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+        .then((response) => {
+          response.interactions.map(
+            (interaction: Interaction, index: integer, array: Interaction[]) => {
+              if (index === array.length - 1) {
+                setConversationId(interaction.conversation_id);
+              }
+            }
+          );
+
+          response.messages.map((message: { type: string; content: string }) => {
```
Comment (wanglam) [inline public/components/incontext_insight/generate_popover_body.tsx:71]: Can we refactor with `response.messages.forEach`? `map` always used to mapping a message to some new object.

File: public/components/incontext_insight/generate_popover_body.tsx [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+              type: 'input',
+              content: summarizationQuestion,
+              contentType: 'text',
+              context: { content: contextContent, dataSourceId: incontextInsight.datasourceId },
+              promptPrefix: getAssistantRole(incontextInsight.key),
+            },
+          }),
+        })
+        .then((response) => {
+          response.interactions.map(
```
Comment (wanglam) [inline public/components/incontext_insight/generate_popover_body.tsx:63]: Shall we refactor like below code?
```
if(response.interactions.length > 0){
    setConversationId(response.interactions[response.interactions.length - 1].conversation_id);
}
```

File: public/components/incontext_insight/generate_popover_body.tsx [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+        });
+    };
+
+    return summarize();
+  };
+
+  if (!isLoading)
+    return (
+      <EuiButton
+        onClick={async () => {
```
Comment (wanglam) [inline public/components/incontext_insight/generate_popover_body.tsx:93]: Shall we call `setIsLlmResponded` to set `isLlmResponded` false at beginning.

File: server/services/chat/olly_chat_service.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (wanglam) [reply server/services/chat/olly_chat_service.ts]: It seems still in, `server/types`, shall we move them to `public/utils/constants.ts`?

[the remark this answers is not in the document]
Comment (songkant-aws) [reply server/services/chat/olly_chat_service.ts]: My bad, looked at the wrong path. It's in `public/utils/constants.ts` now.

File: public/components/incontext_insight/generate_popover_body.tsx (hunk 7 above)
Comment (songkant-aws) [reply public/components/incontext_insight/generate_popover_body.tsx]: Done

File: public/components/incontext_insight/generate_popover_body.tsx (hunk 6 above)
Comment (songkant-aws) [reply public/components/incontext_insight/generate_popover_body.tsx]: Done.

File: public/components/incontext_insight/generate_popover_body.tsx (hunk 8 above)
Comment (songkant-aws) [reply public/components/incontext_insight/generate_popover_body.tsx:89]: Added setIsLlmResponded(false) inside onGenerateSummary.

File: public/components/incontext_insight/generate_popover_body.tsx [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+          }
+          setIsLlmResponded(true);
+        })
+        .catch((error) => {
+          toasts.addDanger(
+            i18n.translate('assistantDashboards.incontextInsight.generateSummaryError', {
+              defaultMessage: 'Generate summary error',
+            })
+          );
+          setIsLoading(false);
```
Comment (wanglam) [inline public/components/incontext_insight/generate_popover_body.tsx:79]: Shall we move this `setIsLoading(false)` to the `.finally` section? Seems the `isLoading` will always be `true` if no error throwed.

Comment (songkant-aws) [reply public/components/incontext_insight/generate_popover_body.tsx:80]: Realized I could simplify the states of those transitions. The summary response and isLlmResponded is overlapped. The initial implementation using isLoading && isLlmResponded to determine whether to show the summary text box. But that's not necessary, checking the content of summary can achieve the same effect of isLlmResponded.

There might be some lifecycle problem of component. After closing the popover, the summary content is also cleared(should be expected). Might be due to the component is not rendered anymore and garbage collected? So I don't handle the clear summary and conversationId in onChatContinuation.

Comment (wanglam) [reply public/components/incontext_insight/generate_popover_body.tsx:80]: I guess the `GeneratePopoverBody` may unmount after `closePopover` called. The `isOpen` is a part of render condition in the `OuiPopover`. The `GeneratePopoverBody` won't be render if not open. Here is the detailed code: https://github.com/opensearch-project/oui/blob/main/src/components/popover/popover.tsx#L748

**Review by wanglam — APPROVED** [review]
Great work! LGTM!

**Review by Hailong-am — APPROVED** [review]
seems there has a conflict for changelog file

**Comment by songkant-aws** [comment]
> seems there has a conflict for changelog file

Resolved changelog conflict

**Comment by raintygao** [comment]
> > seems there has a conflict for changelog file
> 
> Resolved changelog conflict

@songkant-aws  Should this PR be merged in 2.16? If so, I suppose we should add a label and update release notes.

**Review by raintygao — APPROVED** [review]

**Comment by songkant-aws** [comment]
@raintygao As far as I know, it's not required to be merged in 2.16.

**Pull request merged**